[quant] Prep for conv_transpose packing

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43199 [quant] conv_transpose graph patterns
* #40371 [quant] ConvTranspose1d / ConvTranspose2d
* #40370 [quant] conv_transpose1d / conv_transpose2d
* #40360 [quant] conv_transpose1d_prepack / conv_transpose1d_unpack
* #40351 [quant] conv_transpose2d_prepack/conv_transpose2d_unpack
* **#39714 [quant] Prep for conv_transpose packing**



Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

Commit:
[quant] Prep for conv_transpose packing

[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"




[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"




[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"




[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/fbgemm_utils.h:
```diff
@@ -198,6 +210,10 @@ struct CAFFE2_API PackedConvWeight : public ConvPackedParamsBase<kSpatialDim> {
   }
 
+  bool transpose() const override {
+    return (bool)transpose_;
```
why do we need type conversion here?

z-a-f replied:
`transpose_` is `uint8_t` (doesn't work if I use `bool`). Just using explicit conversion

jamesr66a commented on aten/src/ATen/native/quantized/cpu/fbgemm_utils.cpp:
```diff
@@ -212,16 +212,20 @@ Tensor ConvertToChannelsLast3dTensor(const Tensor& src) {
+    torch::List<at::Tensor> /*dilation*/,
+    at::Tensor /*groups*/,
+    at::Tensor /*transpose*/,
+    torch::List<at::Tensor> /*output_padding*/>;
```
This breaks backward compatibility:

```
Jun 11 08:45:59 The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not. 
Jun 11 08:45:59 
Jun 11 08:45:59 Broken ops: [
Jun 11 08:45:59 	__getstate__(__torch__.torch.classes.quantized.Conv2dPackedParamsBase _0) -> ((Tensor, Tensor?, Tensor[], Tensor[], Tensor[], Tensor) _0)
Jun 11 08:45:59 	__setstate__(__torch__.torch.classes.quantized.Conv2dPackedParamsBase _0, (Tensor, Tensor?, Tensor[], Tensor[], Tensor[], Tensor) _1) -> (None _0)
Jun 11 08:45:59 	__getstate__(__torch__.torch.classes.quantized.Conv3dPackedParamsBase _0) -> ((Tensor, Tensor?, Tensor[], Tensor[], Tensor[], Tensor) _0)
Jun 11 08:45:59 	__setstate__(__torch__.torch.classes.quantized.Conv3dPackedParamsBase _0, (Tensor, Tensor?, Tensor[], Tensor[], Tensor[], Tensor) _1) -> (None _0)
Jun 11 08:45:59 ]
```

Unfortunately I don't think there's going to be an easy way around this. What you might have to do is re-bind ConvPackedParams as a new type. When you do that, I suggest making the `SerializationType` generic like in packed RNN cells: https://github.com/pytorch/pytorch/blob/eace0533985641d9c2f36e43e3de694aca886bd9/aten/src/ATen/native/RNN.cpp#L52

That way you can change the specific data that's encoded freely

z-a-f replied:
Any idea how to repro this locally? I am running `python test/backward_compatibility/check_backward_compatibility.py --new-schemas ~/Desktop/new_schemas.txt`, with `new_schemas.txt` created by the `dump_all_function_schemas.py`. But when I run it locally, it just says `Found backward compatible schemas for all existing schemas`

z-a-f replied:
Temporarily resolving it -- will be fixed when #40003 is updated and landed

Commit:
Update on "[quant] Prep for conv_transpose packing"




[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"


Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"


Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"


Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"




Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"




Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"




Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"




Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"




Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"




Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"




Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"




Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"




Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"




Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"




Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"




Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

vkuzo commented on aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp:
```diff
@@ -45,7 +47,8 @@ c10::intrusive_ptr<ConvPackedParamsBase<kSpatialDim>> PackedConvWeight<
       kSpatialDim,
       "D convolution.");
-  const int output_channels = weight.size(0);
+  const int output_channels_idx = transpose ? 1 : 0;
```
optional: if you want these to be more readable, you could define an enum `ConvType` which can be `ConvType::CONV | ConvType::CONV_TRANSPOSE` and pass that around instead

z-a-f replied:
I love this solution -- I was originally thinking of casting boolean to int (as in `int(transpose)`), but thought that would be confusing when used as a channel index

vkuzo commented on aten/src/ATen/native/quantized/cpu/qnnpack_utils.h:
```diff
@@ -114,11 +118,13 @@ struct PackedConvWeightsQnnp : public ConvPackedParamsBase<kSpatialDim> {
             {(uint32_t)dilation_[1], (uint32_t)dilation_[0]},
             {(uint32_t)padding_[0], (uint32_t)padding_[1],
              (uint32_t)padding_[0], (uint32_t)padding_[1]},
-            /*adjustment=*/{0, 0},
```
what does "adjustment" mean here / is it ok to replace with output padding?

z-a-f replied:
Yea, the terminology sucks -- technically, this is adjustment to the output size, which is the same as the output_padding. The only difference is that adjustment refers to the size in general, while padding refers to the padding to every side separately.

Basically, I am using `asjustment` here because the qnnpack uses this as terminology, while pytorch uses "padding" -- should I change it?

vkuzo reviewed (commented):
looks reasonable, ideally we could review this for real when the serialization story is agreed upon

Commit:
Update on "[quant] Prep for conv_transpose packing"




Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"




Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"




Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"




Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"




Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"




Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"




Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"




Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"




Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"




Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"




Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"




Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

Commit:
Update on "[quant] Prep for conv_transpose packing"




Differential Revision: [D22087071](https://our.internmc.facebook.com/intern/diff/D22087071)

[ghstack-poisoned]

vkuzo reviewed (approved):
lg if CI passes
